quickwit/quickwit-indexing/src/source/queue_sources/message.rs

```diff
@@ -0,0 +1,295 @@
+    pub delivery_attempts: usize,
+
+    /// The first deadline when the message is received. It can be extended later using the ack_id.
+    pub initial_deadline: Instant,
+}
+
+/// The raw messages as received from the queue abstraction
+pub struct RawMessage {
+    pub metadata: MessageMetadata,
+    pub payload: OwnedBytes,
+}
+
+impl fmt::Debug for RawMessage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RawMessage")
+            .field("metadata", &self.metadata)
+            .field("payload", &"<bytes>")
+            .finish()
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum PreProcessingError {
+    /// The message can be acknowledged without processing
+    #[error("skippable message: {reason}")]
+    Discardable {
+        reason: &'static str,
+        ack_id: String,
+    },
+    #[error("unexpected message format: {0}")]
+    UnexpectedFormat(#[from] anyhow::Error),
+}
+
+impl RawMessage {
+    pub fn pre_process(
+        self,
+        message_type: MessageType,
+    ) -> Result<PreProcessedMessage, PreProcessingError> {
+        let payload = match message_type {
+            MessageType::S3Notification => PreProcessedPayload::ObjectUri(
+                uri_from_s3_notification(&self.payload, &self.metadata.ack_id)?,
+            ),
+            MessageType::RawUri => {
+                let payload_str = read_to_string(self.payload).context("failed to read payload")?;
+                PreProcessedPayload::ObjectUri(Uri::from_str(&payload_str)?)
+            }
+        };
+        Ok(PreProcessedMessage {
+            metadata: self.metadata,
+            payload,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum PreProcessedPayload {
+    ObjectUri(Uri),
+}
+
+impl PreProcessedPayload {
+    pub fn partition_id(&self) -> PartitionId {
+        match &self {
+            Self::ObjectUri(uri) => PartitionId::from(uri.as_str()),
+        }
+    }
+}
+
+/// A message that went through the minimal transformation to discover its
+/// partition id. Indeed, the message might be discarded if the partition was
+/// already processed, so it's better to avoid doing unnecessary work at this
+/// stage.
+#[derive(Debug, PartialEq, Eq)]
+pub struct PreProcessedMessage {
+    pub metadata: MessageMetadata,
+    pub payload: PreProcessedPayload,
+}
+
+impl PreProcessedMessage {
+    pub fn partition_id(&self) -> PartitionId {
+        self.payload.partition_id()
+    }
+}
+
+fn uri_from_s3_notification(message: &OwnedBytes, ack_id: &str) -> Result<Uri, PreProcessingError> {
+    let value: Value =
+        serde_json::from_slice(message.as_slice()).context("invalid JSON message")?;
+    if matches!(value["Event"].as_str(), Some("s3:TestEvent")) {
+        return Err(PreProcessingError::Discardable {
+            reason: "S3 test event",
+            ack_id: ack_id.to_string(),
+        });
+    }
+    let key = value["Records"][0]["s3"]["object"]["key"]
```

Comment:
What if the event is not a PUT? 
I think we need to check the event type.

Reply:
This seems needlessly rigid but ok.

Reply:
You are right, this is excessively rigid, these can be acknowledged. I'll leave a warn! because this user configuration generates unnecessary traffic on the queue.

Reply:
ok